modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/transaction/ExtensionTransaction.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.transaction;
+
+import static java.lang.String.format;
+import static java.util.Optional.empty;
+import static org.mule.runtime.core.config.i18n.CoreMessages.transactionCanOnlyBindToResources;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.transaction.TransactionException;
+import org.mule.runtime.core.transaction.AbstractSingleResourceTransaction;
+import org.mule.runtime.core.transaction.IllegalTransactionStateException;
+
+import java.util.Optional;
+
+/**
+ * Generic single resource transaction for Extensions
+ *
+ * @since 4.0
+ */
+//TODO: MULE-8946 TransactionalConnection should throw TransactionException
+public class ExtensionTransaction extends AbstractSingleResourceTransaction
+{
+
+    private Optional<ExtensionTransactionalResource> boundResource = empty();
+
+    /**
+     * {@inheritDoc}
+     */
+    public ExtensionTransaction(MuleContext muleContext)
+    {
+        super(muleContext);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void bindResource(Object key, Object resource) throws TransactionException
+    {
+        if (!(key instanceof ExtensionTransactionKey) || !(resource instanceof ExtensionTransactionalResource))
+        {
+            throw new IllegalTransactionStateException(transactionCanOnlyBindToResources(format(
+                    "%s/%s", ExtensionTransactionKey.class.getName(), ExtensionTransactionalResource.class.getName())));
+        }
+
+        ExtensionTransactionalResource txResource = (ExtensionTransactionalResource) resource;
+
+        boundResource = Optional.of(txResource);
+        doBegin();
```

Comment:
Should not call doBegin after super.bindResource was called? (seems to me that it could produce an inconsistent state if the parent call fails)